soc/arm/nxp_kinetis/ke1xf/power.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2021 Vestas Wind Systems A/S
+ * Copyright (c) 2021, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <logging/log.h>
+#include <pm/pm.h>
+#include <soc.h>
+
+LOG_MODULE_DECLARE(power, CONFIG_PM_LOG_LEVEL);
+
+#ifdef CONFIG_XIP
+__ramfunc static void wait_for_flash_prefetch_and_idle(void)
+{
+	uint32_t i;
+
+	for (i = 0; i < 8; i++) {
+		arch_nop();
+	}
+
+	k_cpu_idle();
+}
+#endif /* CONFIG_XIP */
+
+void pm_power_state_set(struct pm_state_info info)
+{
+	switch (info.state) {
+	case PM_STATE_RUNTIME_IDLE:
+		k_cpu_idle();
+		break;
+	case PM_STATE_SUSPEND_TO_IDLE:
+		/* Set partial stop mode and enable deep sleep */
+		SMC->STOPCTRL = SMC_STOPCTRL_PSTOPO(info.substate_id);
+		SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+#ifdef CONFIG_XIP
+		wait_for_flash_prefetch_and_idle();
```

Comment:
Why did you include idle in this function?

Reply:
Do you mean why I included "idle" in the name of the function - or why I included the call to `k_cpu_idle()` in the function?

Reply:
No. The call to `k_cpu_idle()` needs to be executed from RAM for entering STOP mode, so it needs to be in the `__ramfunc` function. Otherwise we will trigger another flash prefetch?

This is similar to `s_stopRamFuncArray` code in the NXP HAL file `mcux/drivers/kinetis/fsl_smc.c`, by the way.